src/main/java/org/sonarsource/sonarlint/ls/AnalysisManager.java

```diff
@@ -313,25 +316,42 @@ private IssueListener createIssueListener() {
       ClientInputFile inputFile = issue.getInputFile();
       if (inputFile != null) {
         URI uri = inputFile.getClientObject();
-        issuesPerFileURI.computeIfAbsent(uri, u -> new ArrayList<>()).add(issue);
+        issuesPerIdPerFileURI.computeIfAbsent(uri, u -> new HashMap<>()).put(UUID.randomUUID().toString(), issue);
       }
     };
   }
 
   Optional<Issue> getIssueForDiagnostic(URI fileUri, Diagnostic d) {
-    return issuesPerFileURI.getOrDefault(fileUri, Collections.emptyList())
-      .stream()
-      .filter(i -> i.getRuleKey().equals(d.getCode().getLeft()) && locationMatches(i, d))
-      .findFirst();
+    Map<String, Issue> issuesForFile = issuesPerIdPerFileURI.getOrDefault(fileUri, emptyMap());
+    String issueKey = Optional.ofNullable(d.getData())
+      .map(JsonPrimitive.class::cast)
+      .map(JsonPrimitive::getAsString)
+      .orElse(null);
+    if (issuesForFile.containsKey(issueKey)) {
+      return Optional.of(issuesForFile.get(issueKey));
+    } else {
+      return issuesForFile.values()
```

Comment:
Fallback to old-style, best-effort matching if `data` is not passed.

Reply:
Why would data not being passed? Old VSCode versions?

Reply:
Yup, though this should be mitigated by the bump in the new minimal version (1.52).